### PR TITLE
🐛 Fix `alias` support for Pydantic v2

### LINKED ISSUE #1536: `Field(alias="")` doesn't work with Pydantic V2
### Privileged issue

- [x] I'm @tiangolo or he asked me directly to create an issue here.

### Issue Content

Currently `alias` parameter doesn't work as expected with Pydantic V2 installed.
Let's use this issue to track this.

👇 Here is a list of tests that showcase the problem (**in the details**):

<details>

```py
from typing import Type, Union

import pytest
from pydantic import VERSION, BaseModel, ValidationError
from pydantic import Field as PField
from sqlmodel import Field, SQLModel

# -----------------------------------------------------------------------------------
# Models


class PydanticUser(BaseModel):
    full_name: str = PField(alias="fullName")


class SQLModelUser(SQLModel):
    full_name: str = Field(alias="fullName")


# Models with config (validate_by_name=True)


if VERSION.startswith("2."):

    class PydanticUserWithConfig(PydanticUser):
        model_config = {"validate_by_name": True}

    class SQLModelUserWithConfig(SQLModelUser):
        model_config = {"validate_by_name": True}

else:

    class PydanticUserWithConfig(PydanticUser):
        class Config:
            allow_population_by_field_name = True

    class SQLModelUserWithConfig(SQLModelUser):
        class Config:
            allow_population_by_field_name = True


# -----------------------------------------------------------------------------------
# Tests

# Test validate by name


@pytest.mark.parametrize("model", [PydanticUser, SQLModelUser])
def test_create_with_field_name(model: Union[Type[PydanticUser], Type[SQLModelUser]]):
    with pytest.raises(ValidationError):
        model(full_name="Alice")


@pytest.mark.parametrize("model", [PydanticUserWithConfig, SQLModelUserWithConfig])
def test_create_with_field_name_with_config(
    model: Union[Type[PydanticUserWithConfig], Type[SQLModelUserWithConfig]],
):
    user = model(full_name="Alice")
    assert user.full_name == "Alice"


# Test validate by alias


@pytest.mark.parametrize(
    "model",
    [PydanticUser, SQLModelUser, PydanticUserWithConfig, SQLModelUserWithConfig],
)
def test_create_with_alias(
    model: Union[
        Type[PydanticUser],
        Type[SQLModelUser],
        Type[PydanticUserWithConfig],
        Type[SQLModelUserWithConfig],
    ],
):
    user = model(fullName="Bob")  # using alias
    assert user.full_name == "Bob"


# Test validate by name and alias


@pytest.mark.parametrize("model", [PydanticUserWithConfig, SQLModelUserWithConfig])
def test_create_with_both_prefers_alias(
    model: Union[Type[PydanticUserWithConfig], Type[SQLModelUserWithConfig]],
):
    user = model(full_name="IGNORED", fullName="Charlie")
    assert user.full_name == "Charlie"  # alias should take precedence


# Test serialize


@pytest.mark.parametrize("model", [PydanticUser, SQLModelUser])
def test_dict_default_uses_field_names(
    model: Union[Type[PydanticUser], Type[SQLModelUser]],
):
    user = model(fullName="Dana")
    data = user.dict()
    assert "full_name" in data
    assert "fullName" not in data
    assert data["full_name"] == "Dana"


# Test serialize by alias


@pytest.mark.parametrize("model", [PydanticUser, SQLModelUser])
def test_dict_default_uses_aliases(
    model: Union[Type[PydanticUser], Type[SQLModelUser]],
):
    user = model(fullName="Dana")
    data = user.dict(by_alias=True)
    assert "fullName" in data
    assert "full_name" not in data
    assert data["fullName"] == "Dana"


# Test json by alias


@pytest.mark.parametrize("model", [PydanticUser, SQLModelUser])
def test_json_by_alias(
    model: Union[Type[PydanticUser], Type[SQLModelUser]],
):
    user = model(fullName="Frank")
    json_data = user.json(by_alias=True)
    assert ('"fullName":"Frank"' in json_data) or ('"fullName": "Frank"' in json_data)
    assert "full_name" not in json_data
```

When run with Pydantic V2:
```
t.py::test_create_with_field_name[PydanticUser] PASSED
t.py::test_create_with_field_name[SQLModelUser] FAILED
t.py::test_create_with_field_name_with_config[PydanticUserWithConfig] PASSED
t.py::test_create_with_field_name_with_config[SQLModelUserWithConfig] PASSED
t.py::test_create_with_alias[PydanticUser] PASSED
t.py::test_create_with_alias[SQLModelUser] FAILED
t.py::test_create_with_alias[PydanticUserWithConfig] PASSED
t.py::test_create_with_alias[SQLModelUserWithConfig] FAILED
t.py::test_create_with_both_prefers_alias[PydanticUserWithConfig] PASSED
t.py::test_create_with_both_prefers_alias[SQLModelUserWithConfig] FAILED
t.py::test_dict_default_uses_field_names[PydanticUser] PASSED
t.py::test_dict_default_uses_field_names[SQLModelUser] FAILED
t.py::test_dict_default_uses_aliases[PydanticUser] PASSED
t.py::test_dict_default_uses_aliases[SQLModelUser] FAILED
t.py::test_json_by_alias[PydanticUser] PASSED
t.py::test_json_by_alias[SQLModelUser] FAILED
```

</details>

All tests parameterized with both, Pydantic model and SQLModel model, to show the difference in behavior.

With Pydantic V1 all tests pass.
With Pydantic V2 most of tests fail with SQLModel model.

I would also expect that `alias` is used as a default value of column name. But it should be possible to override it by providing `sa_column=Column("column_name")` or `sa_column_kwargs={"name": "column_name"}`.

Any feedback is welcome. Please, review the set of tests, suggest use-cases that are currently not covered.

Would be nice to add set of tests for `validation_alias` and `serialization_alias` as well.

### PR DESCRIPTION
**Pull request by ravishan16**
## Description

This PR fixes the Field(alias="...") functionality that was broken in Pydantic v2 while maintaining full backward compatibility with Pydantic v1.

## Problem

In Pydantic v2, the `alias` parameter in SQLModel's `Field` function was not working correctly for validation and serialization. The issue was that Pydantic v2 requires explicit `validation_alias` and `serialization_alias` parameters, but SQLModel wasn't automatically propagating the `alias` value to these parameters.

### Changes Made:

1. **Enhanced Field function**:
   - Added `validation_alias` and `serialization_alias` parameters
   - Implemented version-specific parameter handling

2. **Automatic alias propagation**:
   - ~~Modified `post_init_field_info` in `_compat.py` to automatically set `validation_alias` and `serialization_alias` from `alias` when not explicitly provided (Pydantic v2 only)~~ UPD YuriiMotov: updated it to just using `alias` value for `validation_alias` and `serialization_alias` if they are not specified directly

3. **Backward compatibility**:
   - Filter out v2-only parameters from FieldInfo in Pydantic v1
   - Raise clear RuntimeError when v2 features are used with v1
   - Conditional test execution based on Pydantic version

4. **Comprehensive testing**:
   - Added extensive test coverage for both Pydantic v1 and v2
   - Tests automatically skip v2-specific features when running with v1
   - Covers all alias scenarios including edge cases
   
Breaking Changes
None. This is purely a bug fix with backward compatibility maintained.

Closes
Fixes #1536

### CODE REVIEW

File: sqlmodel/_compat.py [hunk 1]
```diff
@@ -221,7 +221,13 @@ def get_field_metadata(field: Any) -> Any:
         return FakeMetadata()
 
     def post_init_field_info(field_info: FieldInfo) -> None:
-        return None
+        if IS_PYDANTIC_V2:
```
Comment (YuriiMotov) [inline sqlmodel/_compat.py]: This function is declared inside the `if IS_PYDANTIC_V2:` block. So, this condition looks not necessary here.
Am I missing something?

File: sqlmodel/main.py [hunk 2]
```diff
@@ -387,43 +395,65 @@ def Field(
...
+            {
+                "validation_alias": validation_alias,
+                "serialization_alias": serialization_alias,
+            }
+        )
+        field_info = FieldInfo(
+            default,
+            default_factory=default_factory,
+            **field_info_kwargs,
+        )
```
Comment (YuriiMotov) [inline sqlmodel/main.py]: Looks like we can move this outside of `if..else..` block to avoid duplication

File: tests/test_aliases.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+import pytest
+from pydantic import VERSION, BaseModel, ValidationError
+from pydantic import Field as PField
+from sqlmodel import Field, SQLModel
+
+# -----------------------------------------------------------------------------------
+# Models
+
+
+class PydanticUser(BaseModel):
```
Comment (YuriiMotov) [inline tests/test_aliases.py:16]: We should think whether we want to keep parameterizing tests with Pydantic models.
Initially I parametrized tests with both, Pydantic and SQLModel models to show the difference in behavior between SQLModel and Pydantic models.

If it's useful to keep this parametrization in repo? 🤔
Tests would be a bit simpler without it, but on the other hand this way we ensure it works the same as with Pydantic model

File: tests/test_aliases.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
+from typing import Type, Union
+
+import pytest
+from pydantic import VERSION, BaseModel, ValidationError
+from pydantic import Field as PField
+from sqlmodel import Field, SQLModel
+
+# -----------------------------------------------------------------------------------
+# Models
```
Comment (YuriiMotov) [inline tests/test_aliases.py]: We will need to remove unnecessary comments in final version

File: tests/test_aliases.py [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+else:
+    # Dummy classes for Pydantic v1 to prevent import errors
+    PydanticUserV2 = None
+    SQLModelUserV2 = None
+
+
+@pytest.mark.skipif(
+    not VERSION.startswith("2."),
+    reason="validation_alias and serialization_alias are not supported in Pydantic v1",
+)
```
Comment (YuriiMotov) [inline tests/test_aliases.py]:
```suggestion
@needs_pydanticv2
```

We should use `@needs_pydanticv2` decorator as [it's usually done](https://github.com/fastapi/sqlmodel/blob/69b20188bdfec20bcda11c2b73be3e3d571047a8/tests/test_annotated_uuid.py#L9) in this code base
```
from tests.conftest import needs_pydanticv2
```

File: tests/test_aliases.py [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        )
+
+    class SQLModelUserV2(SQLModel):
+        first_name: str = Field(
+            validation_alias="firstName", serialization_alias="f_name"
+        )
+else:
+    # Dummy classes for Pydantic v1 to prevent import errors
+    PydanticUserV2 = None
+    SQLModelUserV2 = None
```
Comment (YuriiMotov) [inline tests/test_aliases.py:137]: We also need to test the logic with `RuntimeError` for Pydantic V1

```py
        if validation_alias:
            raise RuntimeError("validation_alias is not supported in Pydantic v1")
        if serialization_alias:
            raise RuntimeError("serialization_alias is not supported in Pydantic v1")
```

File: sqlmodel/main.py [hunk 7]
```diff
@@ -215,6 +215,8 @@ def Field(
     *,
     default_factory: Optional[NoArgAnyCallable] = None,
     alias: Optional[str] = None,
+    validation_alias: Optional[str] = None,
```
Comment (YuriiMotov) [inline sqlmodel/main.py:221]: Pydantic's `validation_alias` type annotation is wider (`validation_alias: str | AliasPath | AliasChoices | None`), but I think it's fine if we only support `str` for now and extend it later

**Review by YuriiMotov — COMMENTED** [review]
The idea looks good to me!


PS: @svlandeg, sorry, I only noticed you self-assigned this PR when I was about to press "Submit review".. Hope you don't mind I submitted this review)

File: sqlmodel/_compat.py (hunk 1 above)
Comment (ravishan16) [reply sqlmodel/_compat.py]: Added for backward compatibility with Pydantic v1.
Without this, Pydantic tests v1 would fail.

File: sqlmodel/main.py (hunk 2 above)
Comment (ravishan16) [reply sqlmodel/main.py]: Good catch! I've fixed this by moving it outside the if/else block to avoid duplication.

File: tests/test_aliases.py (hunk 6 above)
Comment (ravishan16) [reply tests/test_aliases.py:137]: Added Tests to test the logic with RuntimeError

File: tests/test_aliases.py (hunk 4 above)
Comment (ravishan16) [reply tests/test_aliases.py]: Removed unnecessary comments

File: tests/test_aliases.py (hunk 3 above)
Comment (ravishan16) [reply tests/test_aliases.py:16]: Personally, I think it’s useful to keep the parametrization for now, since it acts as a kind of integration test between the two libraries.

File: sqlmodel/main.py (hunk 7 above)
Comment (ravishan16) [reply sqlmodel/main.py:221]: I don't have a strong preference on this.

File: tests/test_aliases.py (hunk 5 above)
Comment (ravishan16) [reply tests/test_aliases.py]: Addressed the comment

**Comment by svlandeg** [comment]
> Hope you don't mind I submitted this review

Of course not! I tentatively assign PRs when I want to review them in the near future, just to keep track of them, but sometimes other tasks come in between 🙈 Feel free to review whatever you want!

**Comment by ravishan16** [comment]
@YuriiMotov @svlandeg I’ve addressed all the comments from the review. Let me know if there’s anything else you’d like me to focus on.

File: sqlmodel/_compat.py (hunk 1 above)
Comment (YuriiMotov) [reply sqlmodel/_compat.py]: I just checked it with Python 3.11 and Pydantic 1.10.22 - tests passed.
Could you please check it one more time?
```py
    def post_init_field_info(field_info: FieldInfo) -> None:
        if field_info.alias and not field_info.validation_alias:
            field_info.validation_alias = field_info.alias
        if field_info.alias and not field_info.serialization_alias:
            field_info.serialization_alias = field_info.alias
```

File: tests/test_aliases.py [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        pytest.skip("Only relevant for Pydantic v1")
+    with pytest.raises(
+        RuntimeError, match="validation_alias is not supported in Pydantic v1"
+    ):
+        Field(validation_alias="foo")
+
+
+def test_serialization_alias_runtimeerror_pydantic_v1():
+    if VERSION.startswith("2."):
+        pytest.skip("Only relevant for Pydantic v1")
```
Comment (YuriiMotov) [inline tests/test_aliases.py]:
```suggestion
@needs_pydanticv1
def test_serialization_alias_runtimeerror_pydantic_v1():
```

File: tests/test_aliases.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        )
+else:
+    # Dummy classes for Pydantic v1 to prevent import errors
+    PydanticUserV2 = None
+    SQLModelUserV2 = None
+
+
+def test_validation_alias_runtimeerror_pydantic_v1():
+    if VERSION.startswith("2."):
+        pytest.skip("Only relevant for Pydantic v1")
```
Comment (YuriiMotov) [inline tests/test_aliases.py]:
```suggestion
@needs_pydanticv1
def test_validation_alias_runtimeerror_pydantic_v1():
```

File: tests/test_aliases.py [hunk 10]
```diff
@@ -0,0 +1,178 @@
+from typing import Type, Union
+
+import pytest
+from pydantic import VERSION, BaseModel, ValidationError
+from pydantic import Field as PField
+from sqlmodel import Field, SQLModel
+
+from tests.conftest import needs_pydanticv2
```
Comment (YuriiMotov) [inline tests/test_aliases.py]:
```suggestion
from tests.conftest import needs_pydanticv1, needs_pydanticv2
```

**Review by YuriiMotov — COMMENTED** [review]
Good! We are almost there)

I added a few comments.

Tested it a bit more an found a use case that is broken with this PR applied.
This is the workaround people use for now. Although it's not documented and it's a misuse of `schema_extra`, I think it would be nice to support passing `validation_alias` and `serialization_alias` via `schema_extra`, and show a warning message in such cases.
```py
from sqlmodel import SQLModel, Field

class MyModel(SQLModel):
    f: str = Field(schema_extra={"validation_alias": "f_alias"})

m = MyModel.model_validate({"f_alias": "asd"})
assert m.f == "asd"
```

Also, we may want to add more tests (just ideas):
* `alias` + `validation_alias` = `validation_alias` takes precedence
* `alias` + `serialization_alias` = `serialization_alias` takes precedence
* `alias_generator` in model config
* `alias_generator` in model config + `alias` for field = `alias` takes precedence
* passing `validation_alias` and `serialization_alias` via `schema_extra`

File: tests/test_aliases.py (hunk 8 above)
Comment (ravishan16) [reply tests/test_aliases.py]: Thanks fixed it.

File: tests/test_aliases.py (hunk 9 above)
Comment (ravishan16) [reply tests/test_aliases.py]: Thanks fixed it.

File: sqlmodel/_compat.py (hunk 1 above)
Comment (ravishan16) [reply sqlmodel/_compat.py]: You are right. I checked and tests passed. Thanks.

**Comment by ravishan16** [comment]
@YuriiMotov I have made changes based on your suggestion.

schema_extra. Precedence is: explicit params > schema_extra > alias propagation. In v1, using these keys raises a RuntimeError. No warnings added.

Alias generator: Added tests confirming alias_generator is supported in Pydantic v2 for SQLModel. 

Explicit field alias still takes precedence over the generator for both input and output. 

Tests - All passing.

Let me know if this looks ok.

**Comment by ravishan16** [comment]
> The idea looks good to me!
> 
> A couple of more things:
> 
> * I think it would be nice to use `alias` as a default value for DB column. It can then be overridden by `sa_column=Column("column_name")`
> * We should probably test how it works with alias generators specified in model's config (is it supported? - UPD: YES)
> 
> PS: @svlandeg, sorry, I only noticed you self-assigned this PR when I was about to press "Submit review".. Hope you don't mind I submitted this review)

I think it would be better to keep this PR focused on Pydantic v2 alias support, including schema_extra. I suggest opening a follow-up issue to handle using the DB column alias as the default. Let me know your thoughts.

**Comment by YuriiMotov** [comment]
I simplified code a bit - moved alias propagation logic to `Field.__init__` so that it would be all in one place.
Also, got rid of deprecation warnings in test.

**Comment by xdewx** [comment]
Could you please let me know when this PR will be merged and released?

**Review by tiangolo — APPROVED** [review]
Awesome, thank you @ravishan16! 🚀 

And thanks @YuriiMotov! 🍰 

This will be available in SQLModel 0.0.29, released in the next hours. 🎉
